### PR TITLE
Move Chronos to Marathon

### PR DESCRIPTION
**Pull request by dmitriybondr**
https://github.com/CiscoCloud/microservices-infrastructure/issues/792
I made some changes due to this issue.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,4 +8,5 @@ ssl/serial*
 
 # terraform
 **/*.tfstate*
-**/.terraform
\ No newline at end of file
+**/.terraform
+.idea/
```
Comment (corebug) [inline .gitignore]: `.idea/` shouldn't be here, we're trying to keep such stuff out of project's `.gitignore`.

File: roles/glusterfs/tasks/main.yml [hunk 2]
```diff
@@ -4,7 +4,7 @@
   get_url:
     url: "{{ glusterfs_repo }}"
     dest: /etc/yum.repos.d/glusterfs-epel.repo
-    sha256sum: "{{ glusterfs_repo_sha256sum }}"
```
Comment (corebug) [inline roles/glusterfs/tasks/main.yml]: If `sha256sum` variable is defined then deployment fails (probably, repo's SHA256 sum changed).

File: roles/marathon_chronos/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+chronos_http_port: 14400
+chronos_mesos_framework_name: chronos
+chronos_cluster_name: mantl
+
+chronos_mesos_task_cpu: 0.1
+chronos_mesos_task_disk: 256
+chronos_mesos_zk_timeout: 10000
+
+#extra options
+chronos.assets_path:
```
Comment (corebug) [inline roles/marathon_chronos/defaults/main.yml]: Do we *really* need such a list of empty variables?
I guess, those can be removed by now or, at least, commented out here and in `roles/marathon_chronos/templates/chronos-marathon.j2`.

@stevendborrelli what do you think?

**Comment by stevendborrelli** [comment]
Adding in @ryane. 

Should we just use the mesosphere package for chronos and launch it via mantl-api?

**Comment by ryane** [comment]
I tend to think so. We are doing some duplicate work by using ansible to run frameworks on marathon + it seems a little less modular. I think this applies to the elasticsearch framework too

**Comment by corebug** [comment]
@ryane i guess it's not completely correct since `Chronos` is supposed to be **core** part of Mantl and actually requires way more actions then just running the framework. I guess we can refactor the code to use Mantl-API for framework installation, but we also have to deal with proxying, healthchecks and so on.
How all those should be deployed with Mantl-API?

**Comment by dmitriybondr** [comment]
@ryane I think it is nice idea, but I need more information about deploying process.
I can install chronos by this command:
curl -X POST -d "{\"name\": \"chronos\"}" http://mantl-worker-003:4001/1/packages
but I don't understand how to specify parameters for marthon.
I found something in consul KV by this path MANTL-INSTALL/REPOSITORY/0/REPO/PACKAGES/C/CHRONOS/0/.
is this what I am looking for ?
if it so, could you please describe how to configure it in a right way?

**Comment by ryane** [comment]
@dmitriybondr this currently doesn't build for me on vagrant:

```
=> default: failed: [default] => {"attempts": 10, "changed": true, "cmd": "curl marathon.service.consul:8080/v2/apps/chronos/tasks | awk ' {print $3} '", "delta": "0:00:00.157471", "end": "2015-12-09 12:46:15.368143", "failed": true, "rc": 0, "start": "2015-12-09 12:46:15.210672", "warnings": ["Consider using get_url module rather than running curl"]}
==> default: stderr:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
==> default:
==> default:            Dload  Upload   Total   Spent    Left  Speed
0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: marathon.service.consul; Name or service not known
```

The reason your Mantl API command failed was because we had not created a support chronos package yet. There are a few issues with the Chronos package PR that need to be addressed but we should have a working package very soon.


@corebug you're right. if we move chronos to marathon, we need to maintain (or improve) the integration we already have. One thing we still need to do before we can make this switch is to update the UI to properly reflect a valid chronos health check (see CiscoCloud/mantl-ui-frontend#4).

**Comment by ryane** [comment]
there is now a working chronos package as of CiscoCloud/mantl-universe#15

File: terraform.sample.yml [hunk 4]
```diff
@@ -115,3 +114,5 @@
     glusterfs_mode: client
   roles:
     - glusterfs
+    - marathon_chronos
```
Comment (ryane) [inline terraform.sample.yml]: can we just continue to use `chronos` as the role name?

File: roles/mesos/defaults/main.yml [hunk 5]
```diff
@@ -11,7 +11,7 @@ mesos_follower_port: 5051
 mesos_leader_cmd: mesos-master
 mesos_follower_cmd: mesos-slave
 mesos_executor_registration_timeout: 10mins
-mesos_resources: "ports(*):[4000-5000, 7000-8000, 9000-10000, 25000-26000, 31000-32000]"
+mesos_resources: "ports(*):[4000-5000, 7000-8000, 9000-10000,14400-14402, 25000-26000, 31000-32000]"
```
Comment (ryane) [inline roles/mesos/defaults/main.yml]: I don't think this is necessary if we switch to the chronos package

File: roles/marathon_chronos/templates/chronos-consul.j2 [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
+ "service": {
+   "name": "chronos",
+   "tags": [ "chronos" ],
+   "port": 14400,
+   "check": {
+     "script": "curl --silent --show-error --fail --dump-header /dev/stderr --retry 2 --connect-timeout 5 --max-time 5 http://chronos.service.consul:14400/ping",
```
Comment (ryane) [inline roles/marathon_chronos/templates/chronos-consul.j2]: Using the chronos.zookeeper.consul dns name in the consul health check might be a problem.

Also, with the packaged version of chronos, creating a consul health check is a bigger challenge. We are looking at just updating the UI to take advantage of the marathon health check instead (CiscoCloud/mantl-ui-frontend#4)

**Comment by ryane** [comment]
the chronos package required a few small changes to mantl-api so I bumped the version number. I know this PR will likely change a lot to use the package but I added some additional comments to what is already here.

File: terraform.sample.yml (hunk 4 above)
Comment (dmitriybondr) [inline terraform.sample.yml]: Of course, it  would be even better.

File: roles/marathon_chronos/templates/chronos-consul.j2 (hunk 6 above)
Comment (dmitriybondr) [inline roles/marathon_chronos/templates/chronos-consul.j2]: I've been investigating consul health checks... I totally agree. updating the UI would be simplier and more logical

**Comment by dmitriybondr** [comment]
@ryane  I found unexpected behaviour, trying to run chronos. I got this error:
"[scallop] Error: Bad arguments for option 'cassandra ttl': '3.1536e+07' - wrong arguments format "
I think it is because mantl-api having faced with some large integers tries to convert it to exponential format.

**Comment by ryane** [comment]
@dmitriybondr this should be fixed in mantl-api 0.1.2 (see 5925df9da5aa69cd97651e58f15be0961d91bfa8 in this PR). are you still seeing that with 0.1.2?

**Comment by dmitriybondr** [comment]
@ryane I still see this this with new version of mantl-api (0.1.2)

**Comment by ryane** [comment]
@dmitriybondr that's strange. I cannot repro the error with 0.1.2 (but I can with 0.1.1). Do you want to try and setup a time where we can go over your setup?

**Comment by dmitriybondr** [comment]
@ryane  I decided to  destroy my cluster and rebuild it from scratch with new mantl-api. And it works ! it awesome. Thank you. )

**Comment by siddharthist** [comment]
@dmitriybondr What's the status of this PR?

**Comment by siddharthist** [comment]
I'm going to close this for now since there hasn't been a commit since December. Please reopen if there's more interest, or we can create an issue tracking the movement of Chronos to Marathon.

**Pull request closed without merging**